### PR TITLE
base/*: fix uncompressed inline child resource with compressed parent

### PR DESCRIPTION
**Pull request by bgilbert**
When desugaring inline/local resource contents, the current rules are:

1. If the input config specifies `compression: gzip`, we pass through the `compression` field unchanged and refuse to compress again.  This is useful for local resources that are stored precompressed alongside the config.
2. If we decide to compress, we set the `compression` field to `gzip`.
3. If we decide not to compress, we do not set the `compression` field, ostensibly so precompressed resources in child configs can inherit the `compression` field from the parent.

But case 3 doesn't really work.  Butane doesn't know that the child resource is precompressed, so is allowed to compress it again, and only the compressibility heuristics (which technically are not contractual) prevent it from doing so and producing invalid output.  (There's no way to represent nested compression, so the file would be written to the provisioned system compressed.)  And if the parent resource is not formally precompressed, but is actually just auto-compressed by Butane, a change to the parent's contents or to the Butane heuristics could cause the parent to become uncompressed in the future and break the merged config.

And case 3 creates a problem for an inline/local child resource overriding an inline/local parent.  If Butane opts to compress the parent but not the child, the merged config will have uncompressed contents but inherit `compression: gzip` from the parent, causing a provisioning failure.  And if the parent/child contents change over time, this problem may suddenly appear in the future.

Fix this by changing rule 3.  When desugaring a non-precompressed inline/local resource, always set the `compression` field to either "" or `gzip`.  This intuitively makes sense: only we know whether we decided to compress the resource, so we need to say so.

This involves an API change to `MakeDataURL()`.  We now return a pointer to the selected `compression` value, or `nil` if we're passing through the `compression` value from the input (case 1).  In the latter case, the caller should record a 1:1 translation for the `compression` field, rather than attributing its value to the `inline`/`local` input field.

Fixes https://github.com/coreos/butane/issues/332.

### CODE REVIEW

File: base/util/url.go [hunk 1]
```diff
@@ -53,10 +65,10 @@ func MakeDataURL(contents []byte, currentCompression *string, allowCompression b
 			return
 		}
 		gz := ";base64," + base64.StdEncoding.EncodeToString(buf.Bytes())
-		// Account for space needed by "compression": "gzip".
-		if len(gz)+25 < len(opaque) {
+		// Account for space needed by the "gzip" compression value
+		if len(gz)+4 < len(opaque) {
```
Comment (cgwalters) [inline base/util/url.go:69]: I would make this `len(gz)+len("gzip")` or so...the compiler should constant fold hopefully, and it makes it more robust to future changes.

File: base/util/url.go [hunk 2]
```diff
@@ -24,9 +24,21 @@ import (
...
+	// contents were compressed by the user; we can pick a data URL
+	// encoding but we can't compress again.  In this case we set our
+	// compression return value to null so the caller knows not to
+	// record a translation from input contents to output compression.
+	// In all other cases, we need to explicitly set the compression
+	// field to avoid a child config inheriting an incorrect compression
+	// setting from the parent.
+	if util.NilOrEmpty(currentCompression) {
+		compression = util.StrToPtr("")
+	}
```
Comment (cgwalters) [inline base/util/url.go:40]: I had to read this comment and code block twice.  The comment starts out talking about a condition, then talks about the "all other cases" - and the code implements the "all other cases".  I kept expecting it to implement what the *start* of the comment was describing.

Maybe:

```
If the config has an empty/nil compression field, we need to 
explicitly set the compression field to avoid a child config
inheriting an incorrect compression setting from the parent.

Otherwise, If the config has a non-empty compression field, then the
contents were compressed by the user; we can pick a data URL
encoding but we can't compress again.  In this case we set our
compression return value to null so the caller knows not to
record a translation from input contents to output compression.
```

maybe?

**Review by cgwalters — APPROVED** [review]
I only gave this a somewhat superficial look but seems sane to me!

**Comment by bgilbert** [comment]
Updated.

File: base/v0_2/translate_test.go [hunk 3]
```diff
@@ -196,7 +196,8 @@ func TestTranslateFile(t *testing.T) {
 							},
 						},
 						{
-							Source: util.StrToPtr("data:,file%20contents%0A"),
+							Source:      util.StrToPtr("data:,file%20contents%0A"),
+							Compression: util.StrToPtr(""),
```
Comment (cgwalters) [inline base/v0_2/translate_test.go:200]: Do we actually need to change all the tests to explicitly specify a `""` compression?  Shouldn't it be equivalent to `nil`?

**Review by jlebon — APPROVED** [review]
LGTM overall. Do we need to surface this somehow also in the Ignition spec docs for the Ignition-level version of the config merge issue?

**Comment by cgwalters** [comment]
OK right, while this seems like a useful fix I don't think it will actually help the MCO today, see https://github.com/openshift/machine-config-operator/pull/3126#issuecomment-1113650288
which I guess we knew from https://github.com/coreos/butane/issues/332#issuecomment-1070861759
anyways.

File: base/v0_2/translate_test.go (hunk 3 above)
Comment (bgilbert) [reply base/v0_2/translate_test.go:200]: Right, the whole point of this PR is to emit `""` rather than `nil` because they're semantically distinct for config merging.

**Comment by bgilbert** [comment]
> Do we need to surface this somehow also in the Ignition spec docs for the Ignition-level version of the config merge issue?

We [describe the field merging semantics](https://coreos.github.io/ignition/operator-notes/#child-configs-take-precedence-when-specified) in Ignition operator notes, so I don't think so?  The issue here was that Butane wasn't taking those semantics into account.

File: base/v0_2/translate_test.go (hunk 3 above)
Comment (cgwalters) [reply base/v0_2/translate_test.go:200]: > Right, the whole point of this PR is to emit "" rather than nil because they're semantically distinct for config merging.

Yeah sorry, it is obvious when stated that way.  I don't know why I find this so confusing.

**Comment by bgilbert** [comment]
Filed https://github.com/coreos/ignition/issues/1357 for better config merging docs, including an explicit callout of overriding the `compression` field.
